src/components/structures/MessagePanel.js

```diff
@@ -440,14 +443,32 @@ export default class MessagePanel extends React.Component {
 
         const ret = [];
 
+        const firstShownEventIndex = checkForPreJoinUISI(
```

Comment:
As you hinted in your comment, the `render` function is not the best place for this kind of work, as it will called for every state and prop change of the component. Instead, I would suggest to do this processing only where the set of events has changed and store your result in component state.